wip/hyperstart: Initial cc-shim implementation

amshinde opened the pull request.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+#include <stdint.h>
+#include <inttypes.h>
+
+#define MAX_POLL_FDS 5
```
Why 5? Can you add a comment explaining what this variable is for and why the value was chosen.

amshinde replied:
Frankly I picked this as an approximate number to start with developing. As we are defining things, I see the proxy handling at max four fds - stdin, signal fd, proxy socket fd and an I/O fd sent from the proxy. I can add those as comments and change to 4

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+#include <inttypes.h>
+
+#define MAX_POLL_FDS 5
+#define BUF_SIZE 1024
```
I'd drop this and use the `stdio.h`-defined `BUFSIZ`.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+#define MAX_POLL_FDS 5
+#define BUF_SIZE 1024
+
+#define WINSIZE 11
```
Can you explain this?

amshinde replied:
These are the command codes thats hyperstart uses for commands on the CTL channel. I'll add a comment.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+#define BUF_SIZE 1024
+
+#define WINSIZE 11
+#define KILLCONTAINER 18
```
Again, this needs a comment.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+/* Pipe used for capturing signal occurence */
+int pipe_fd[2];
+
+void add_pollfd(struct pollfd *poll_fds, nfds_t *nfds, int fd,  short events) {
```
You might want to consider adding Doxygen headers.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+
+	flags = fcntl(fd, F_GETFL);
+	if (flags == -1) {
+		err_exit("fcntl-F_GETFL");
```
Add `errno` to error maybe?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+}
+
+char*
+get_hypertart_msg(char *json, int hyper_cmd_type, size_t *len) {
```
This really needs a function header explaining the format of the hyper message.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+}
+
+void
+err_exit(const char *format, ...)
```
Check arg?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+}
+
+void
+set_fd_nonblocking(int fd)
```
Check arg?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+}
+
+void
+set_big_endian_32(uint8_t *buf, uint32_t val)
```
Check args?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+ }
+
+void
+send_proxy_hyper_message(int fd, int hyper_cmd_type, char *json) {
```
Check args?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+}
+
+void
+handle_signals(char *container_id, int outfd) {
```
Check args?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+	while ((nread = read(infd, buf, BUF_SIZE))!= -1) {
+		ret = (int)write(outfd, buf, (size_t)nread);
+		if (ret == -1) {
+			fprintf(stderr, "Error writing to fd : %s\n", strerror(errno));
```
MIght be useful to log `infd` and `outfd` here.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+void add_pollfd(struct pollfd *poll_fds, nfds_t *nfds, int fd,  short events) {
+	struct pollfd pfd = { 0 };
+
+	assert(*nfds < MAX_POLL_FDS);
```
`assert()` or `err_exit()`? I guess this is an indication of an programming rather than a runtime error, but if the check fails, it could be useful to dump `poll_fds`.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+					strerror(errno));
+				continue;
+			}
+			ret = asprintf(&buf, "{\"container_id\":\"%s\", \"row\":\"%d\", \"col\":\"%d\"}",
```
Might be useful to have a debug option so that this function can log the action performed:

- `"killed container %s with signal %d"`
- `"handled SIGWINCH for container %s (row=%d, col=%d)"`

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,292 @@
+		2 =>sockfd
+	*/
+
+	while (1) {
```
The shim needs to shutdown when the container does (returning the containers exit code), but it doesn't look like that is currently handled?

amshinde replied:
Yes, not yet implemented. I still need clarification on how the stdout/err would be sent by the proxy. I'll add a TODO header/file listing the things that need to be done.

amshinde commented:
Pushed all the review changes

jcvenegas commented on shim/README.md:
```diff
@@ -0,0 +1,12 @@
+
+TODO:
+It should also forward any input received from containerd-shim to cc-proxy and 
+forward the stdout/stderr received from the proxy to containerd-shim.
```
so cc-shim will own and manage of init-std* and exec named pipes provided by docker?

amshinde replied:
no cc-shim just has to write to stdout/err. Those get piped to the named pipes by the way things are set up by docker.

jcvenegas replied:
@amshinde ok, I ask because I think that now the way to have tty io wont be on cc-runtime but proxy or shim by handling mutiplexed unix tty unix socket from hyperstart daemon.

But yes, seems more likely to me do it in the proxy

jcvenegas replied:
@dlespiau what is the plan for tty/stdio manage ?
cc-oci-runtime have 3 stdio cases to handle
if ( not config->console ) // not using --console something
        if (detach_mode)
                send io to docker named pipes //send to pipes used by "docker logs container_id"
        else if( attach_mode ) // running standalone no console 
                send io to console.socket 

else have config->console // using --console something
        use the console provided to show stdio

jcvenegas commented on shim/shim.c:
```diff
@@ -0,0 +1,399 @@
+/* These are the command ids used by hyperstart control message format.
+ * cc-shim sends WINSIZE command on receving a SIGWINCH signal and 
+ * a KILLCONTAINER command on receiving any other signal
+ */
```
would be better to have a struct or a enum that maps all hyper commands in a special file like  hyper_agent.c

amshinde replied:
Yes. I was thinking about it myself, it would be easier to sync the enum/struct. 
I'll add a shim.h and add it there.

jcvenegas commented on shim/shim.c:
```diff
@@ -0,0 +1,399 @@
+
+/* globals */
+/* Pipe used for capturing signal occurence */
+int pipe_fd[2];
```
what about a more meaningful variable name signal_pipe_fd?

amshinde commented:
Added option parsing for arguments.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
+				container_id = strdup(optarg);
+				break;
+			case 'p':
+				if ((ret = (int)strtol(optarg, &endptr, 10)) == -1 || *endptr != '\0') {
```
I'd break this line up into separate statements as that would make it clearer.

Also, `strtol()` doesn't return `-1` on error - you need to check for `LONG_MIN` and `LONG_MAX`. However, it would be just as good to set `errno=0` before the call and just check `(!*endptr && !errno)` after it rather than checking `ret` directly.

amshinde replied:
duh! I had made a note of that from the man page but ended up with checking for -1 instead.
I do like the approach of just checking for errno.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
+
+	struct option prog_opts[] = {
+		{"container_id", required_argument, 0, 'c'},
+		{"proxy_sock_fd", required_argument, 0, 'p'},
```
These long option names should use hyphens rather than underscores.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
+void print_usage(char *prog) {
+	printf("%s: Usage\n", prog);
+        printf("  -c,  --container_id   Container id\n");
+        printf("  -p,  --proxy_sock_fd  File descriptor of the socket connected to cc-proxy\n");
```
Need to update long options to use hyphens, not underscores.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
+				break;
+			case 'h':
+				print_usage(argv[0]);
+				exit(0);
```
`exit(EXIT_SUCCESS)` for consistency with the use of `EXIT_FAILURE` below.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
-	char* container_id;
+	char* container_id = NULL;
 	char* proxy_sock_path;
 	struct pollfd poll_fds[MAX_POLL_FDS];
```
I'd be tempted to set `poll_fds[MAX_POLL_FDS] = { -1 };` for symmetry with `proxy_sock_fd`.

jodh-intel commented on shim/shim.c:
```diff
@@ -296,25 +297,67 @@ handle_data(int infd, int outfd)
-		printf("Usage: cc-shim container_id proxy_socket_path");
-		exit(EXIT_FAILURE);
+	if ( !container_id) {
+		err_exit("cc-shim: Missing container id\n");
```
Hard-coding program name - not major, but `print_usage()` is passed the program name, so ideally `err_exit()` would do something similar (or maybe rely on a global `static const char *program_name` set in `main()`.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+		err_exit("Missing proxy I/O file descriptor\n");
+	}
+
+	if (shim.io_seq_no == 0) {
```
Probably worth explaining why `-e` is optional otherwise someone might see that as a bug and try to fix it ;)

amshinde commented:
Update since last review:
Added command line options for proxy I/O fd and sequence numbers.
Added I/O handling that proxy added.
Changed signal handler for signals that need to be forwarded.
Added logging

Pending:
Move to Nonblocking + Buffered writes to proxy ctl and I/O fds.
Tests with mock proxy.
Exit code handling (This may need some work on proxy)

dlespiau commented:
There are a few warnings that make the travis tests fail. You may have fixed those locally, but this PR is failing as is.

dlespiau commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+	add_pollfd(poll_fds, &nfds, STDIN_FILENO, POLLIN | POLLPRI);
+
+	add_pollfd(poll_fds, &nfds, shim.proxy_io_fd, POLLIN | POLLPRI);
+#if 0
```
commented out code :)

dlespiau commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+
+	*stream_len = 0;
+
+	buf = calloc(STREAM_HEADER_SIZE, 1);
```
No need to heap allocate the header, could just be a static variable in this function or use alloca()

dlespiau replied:
Ok, read a bit more the function. Changing this is a bit annoying, better do it when we have things starting to work and unit tests.

dlespiau commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+			free(buf);
+			return NULL;
+		} else if (ret == 0) {
+			/* TODO: handle this scenario - has the proxy gone away or eof received
```
Right so we've discussed this, the exit code isn't properly handled at the moment it seems. I'm happy to still have my lgtm and we get to add support for this later.

dlespiau commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+	} else {
+		//shim_warning("Seq no %"PRIu64 " received from proxy does not match with
+		//		 shim seq %"PRIu64 "\n", seq, shim->io_seq_no);
+		return;
```
in case of unexpected seq number, still need to free(buf)

dlespiau commented:
pull approve didn't seem to catch it so:

lgtm

amshinde commented:
@dlespiau I am going to push the exit code handling shortly.

dlespiau commented:
Just in case you haven't seen:

  - 4 commits missing the signed-off-by tag
  - very interesting warnings from clang

jodh-intel commented on data/hypervisor.args.in:
```diff
@@ -18,7 +18,7 @@ virtio-9p-pci,fsdev=workload9p,mount_tag=rootfs
 local,id=workload9p,path=@WORKLOAD_DIR@,security_model=none
 -smp
-2,sockets=2,cores=1,threads=1
+2,sockets=1,cores=2,threads=1
```
Branch needs a rebase to avoid regressing the fix for #339 here.

jodh-intel replied:
Ah - sorry, it's not your branch that needs to be rebased by the wip/hyperstart one. Shouldn't this PR be for merging the shim into `master` now?

jodh-intel commented on m4/ax_valgrind_check.m4:
```diff
@@ -205,7 +205,7 @@ VALGRIND_LOG_COMPILER = \
 	$(VALGRIND) $(VALGRIND_SUPPRESSIONS) --error-exitcode=1 $(VALGRIND_FLAGS)
 
-define valgrind_tool_rule =
+define valgrind_tool_rule
```
Again, this is fixed in `master` (#359) but not in `wip/hyperstart`.

jodh-intel commented on shim/log.h:
```diff
@@ -0,0 +1,46 @@
+ *
+ *  LOG_EMERG      system is unusable -> 0
+ *  LOG_ALERT      action must be taken immediately -> 1
+ *  LOG_CRIT       critical conditions
```
Looks like this should be `LOG_CRITICAL`.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,621 @@
+/* globals */
+
+/* Pipe used for capturing signal occurence */
+int signal_pipe_fd[2];
```
Initialise both to `-1` just to be super-safe?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+ * \param fd File descriptor to add
+ * \param events Events for the fd that should be polled
+ */
+void add_pollfd(struct pollfd *poll_fds, nfds_t *nfds, int fd,  short events) {
```
Missing param checks?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+ * \return true on success, false otherwise
+ */
+bool
+assign_all_signals(struct sigaction *sa)
```
Missing param check?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+ * \param[out] len Length of the message
+ */
+char*
+get_proxy_ctl_msg(char *json, size_t *len) {
```
Missing param checks?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+ * \param json Json payload
+ */
+void
+send_proxy_hyper_message(int fd, int hyper_cmd_type, char *json) {
```
Missing param checks?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+ * \return newly allocated string on success, else \c NULL.
+ */
+char*
+read_IO_message(struct cc_shim *shim, uint64_t *seq, ssize_t *stream_len) {
```
Missing param checks?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+}
+
+long long
+parse_numeric_option(char *input) {
```
Missing param check?

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+		{ 0, 0, 0, 0},
+	};
+
+	while ((c = getopt_long(argc, argv, "c:p:o:s:e:dh", prog_opts, NULL))!= -1) {
```
Super pedantic I know, but I always sort the option string and the case statements for easier maintenance.

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+		3 =>sockfd
+	*/
+
+	while (1) {
```
Since there is no normal code path to break out of this loop, I think this deserves a comment explaining why.

jodh-intel commented on shim/shim.h:
```diff
@@ -0,0 +1,79 @@
+	bool        exiting;
+};
+
+// Hyperstart control cmd ids : apiversion 4242
```
Maintenance: might be worth adding a comment stating that the order of these entries must not change. Otherwise,  someone might decide later on to sort the list to make it look better ... ;-)

jodh-intel commented on shim/shim.c:
```diff
@@ -0,0 +1,631 @@
+get_proxy_ctl_msg(char *json, size_t *len) {
+	char *proxy_ctl_msg = NULL;
+
+	*len = strlen(json) + 8 + 1;
```
Shouldn't this be `PROXY_CTL_HEADER_SIZE ` rather than the magic `8`?

amshinde commented:
@dlespiau I had rebased against master. The signed-off missing checks were flagged against the merge commits.

amshinde commented:
Closing this PR, opened one to be merged with master.
